### PR TITLE
DM-13536: Use repositories more idiomatically

### PR DESCRIPTION
**Pull request by kfindeisen**
This PR reworks `ap_verify` to give it more centralized control over the repositories it creates while running. The current strategy is to create three parallel repositories in the directory specified by the user, one for science data, one for calibrations, and one for the output of `ap_pipe`. However, the workspace concept provides enough abstraction that this can be easily changed. In addition, while `ap_pipe` currently requires repository URIs, the system can be easily adapted to pass butlers or datarefs instead.

API changes must be syncedwith lsst-dm/ap_pipe#16.

### CODE REVIEW

File: tests/test_workspace.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
+# LSST Data Management System
+# Copyright 2017 LSST Corporation.
```
Comment (jdswinbank) [inline tests/test_workspace.py]: 2018? AURA/LSST? Per https://github.com/lsst/templates/blob/master/project_templates/stack_package/example/python/lsst/__init__.py#L4.

(Also above.)

File: python/lsst/ap/verify/workspace.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        else:
+            os.makedirs(location, stat.S_IRWXU | stat.S_IRGRP | stat.S_IROTH)
+
+        self._location = location
+
+    @property
+    def dataRepo(self):
+        """The URI to a Butler repo for science data (`str`, read-only).
+        """
+        return os.path.join(self._location, 'ingested')
```
Comment (jdswinbank) [inline python/lsst/ap/verify/workspace.py]: (Here and below)

Being picky, if you really mean URI, is it a good idea to use `os.sep` to join? I imagine it's totally fine on any LSST-supported platfom, but maybe it should really be `urlparse.urljoin`?

File: doc/lsst.ap.verify/running.rst [hunk 3]
```diff
@@ -34,11 +34,11 @@ Using the :ref:`HiTS 2015 <ap_verify_hits2015-package>` dataset as an example, o
 Here:
 
 * :command:`HiTS2015` is the :ref:`dataset name <ap-verify-dataset-name>`,
-* :command:`workspace/hits/` is the location of the :ref:`Butler repository<command-line-task-data-repo-using-uris>` in which the pipeline will work,
+* :command:`workspace/hits/` is the location where the pipeline will create any :ref:`Butler repositories<command-line-task-data-repo-using-uris>` necessary,
```
[the remark this answers is not in the document]
Comment (kfindeisen) [reply doc/lsst.ap.verify/running.rst]: I meant the dictionary definition of "a place where you can get work done" instead of introducing a new concept. What should I call it, then?

File: python/lsst/ap/verify/ap_verify.py [hunk 4]
```diff
@@ -140,8 +141,8 @@ def _measureFinalProperties(metricsJob, metadata, outputDir, args):
         The Job object to which to add any metric measurements made.
     metadata : `lsst.daf.base.PropertySet`
         The metadata produced by the AP pipeline.
-    outputDir : `str`
-        The location of the final processed data repository.
+    workspace : `lsst.ap.verify.workspace.Workspace`
+        The abstract location containing input and output repositories.
```
[the remark this answers is not in the document]
Comment (kfindeisen) [reply python/lsst/ap/verify/ap_verify.py]: I'll remove it.

[the remark this answers is not in the document]
Comment (kfindeisen) [reply python/lsst/ap/verify/ap_verify.py]: Just emphasizing that the function doesn't take something that can be directly interpreted as, say, a filesystem path.

File: python/lsst/ap/verify/workspace.py (hunk 2 above)
Comment (kfindeisen) [reply python/lsst/ap/verify/workspace.py]: A lot of the existing code assumes these are filesystem paths (see [DM-11482](https://jira.lsstcorp.org/browse/DM-11482)). The problem with just using `urlparse` is that what the Butler calls a URI can really be *either* a URI (by the conventional definition) or a filesystem path, so I'd need to be able to handle both.

TL;DR: I suspect that a dumb replacement would break existing code, so I'm deferring the fix to DM-11482.

File: doc/lsst.ap.verify/running.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (jdswinbank) [reply doc/lsst.ap.verify/running.rst]: I don't have any problem with the name; I just reckon it'd be clearer to say something like “we use the term _workspace_ to refer to the user-specifed directory into which the multiple output repositories generated by ap_verify will be written”, or similar.
